larpandoracontent/LArHelpers/LArMCParticleHelper.cc

```diff
@@ -294,33 +430,58 @@ void LArMCParticleHelper::SelectReconstructableMCParticles(const MCParticleList
 
     // Obtain vector: primary mc particles
     MCParticleVector mcPrimaryVector;
-    LArMCParticleHelper::GetPrimaryMCParticleList(pMCParticleList, mcPrimaryVector);
+
+    if (!testBeamHierarchyMode)
+    {
+        LArMCParticleHelper::GetPrimaryMCParticleList(pMCParticleList, mcPrimaryVector);
+    }
+    else
+    {
+        LArMCParticleHelper::GetLeadingMCParticleList(pMCParticleList, mcPrimaryVector);
+    }
 
     // Select MCParticles matching criteria
     MCParticleVector candidateTargets;
     LArMCParticleHelper::SelectParticlesMatchingCriteria(mcPrimaryVector, fCriteria, candidateTargets);
 
     // Ensure the MCParticles have enough "good" hits to be reconstructed
     LArMCParticleHelper::SelectParticlesByHitCount(candidateTargets, mcToTrueHitListMap, mcToPrimaryMCMap, parameters, selectedMCParticlesToHitsMap);
+
+    // ATTN: The parent particle must be in the hierarchy map, event if not reconstructable
+    if (testBeamHierarchyMode)
+    {
+        for (const MCParticle *const pMCParticle : candidateTargets)
+        {
+            if (!LArMCParticleHelper::IsBeamParticle(pMCParticle))
+                continue;
+
+            const MCParticle *const pParentMCParticle(LArMCParticleHelper::GetParentMCParticle(pMCParticle));
+            if (selectedMCParticlesToHitsMap.find(pParentMCParticle) == selectedMCParticlesToHitsMap.end())
+            {
+                CaloHitList caloHitList;
```

Comment:
Can combine these steps using the `operator[]`, which will create a new key, value pair if it doesn't exist, or return a reference to the value if the key is present.